### PR TITLE
refactor(scraper): refactor has_new_grad_signal in update_jobs.py and add more robust tests

### LINKED ISSUE #46: [Beginner]: Add a test case for `has_new_grad_signal` in `tests/test_filter.py`
### Description
The function `has_new_grad_signal(title, signals)` lacks direct unit test coverage. Add a basic suite of tests verifying it identifies new grad keywords correctly.

### Why it matters
Testing matching logic ensures we don't accidentally start filtering out valid junior roles due to regex/string typos.

### Acceptance Criteria
- [ ] Create `test_has_new_grad_signal()` in `tests/test_filter.py`
- [ ] Test with both matching and non-matching titles
- [ ] Passes `pre-commit run --all-files`

### PR DESCRIPTION
**Pull request by half-measures**
## Linked Issue

Fixes #46 

## Summary

In keeping with existing class tests, added 'TestHasNewGradSignal' with multiple tests to test the function of the has_new_grad_signal function located in update_jobs.py.

## Changes Made

FILE                WHAT CHANGED
test_filter.py=Added multiple tests to ensure the test_has_new_grad_signal function is robust and functioning. update_jobs.py=With added tests, to meet other requirements, had to re-work the function to account for more robust requirements of our tests. Had to be able to handle NaN, none and some other edge cases.

## Testing

- [x] `python -m py_compile scripts/update_jobs.py` — no errors
- [x] `make test` — all tests pass
- [x] `pre-commit run --all-files` — clean

## Notes for Reviewer
Had to delete the repo and re clone it as I keep somehow capturing the 2k changes from the main repo and it looks messy. I would assume I need to learn about rebase.
This ended up being a whole thing with re-doing the function to be more stable. 
---

## Summary by CodeRabbit

* **New Features**
  * Improved detection of "new graduate" signals in job listings with more reliable, normalized matching and input validation.

* **Bug Fixes**
  * Reduced false positives; matching is now case-insensitive and correctly handles boundary words, hyphens, Unicode, and special-value/empty-title edge cases.

* **Tests**
  * Added comprehensive unit tests covering matching, non-matching, empty-signal, and many edge-case scenarios.

### CODE REVIEW

File: tests/test_filter.py [hunk 1]
```diff
@@ -140,3 +140,42 @@ def test_different_urls_both_kept(self):
...
+    def test_partial_word_does_not_match(self):
+        """A signal should match a whole word, not a substring of another word."""
+        assert not has_new_grad_signal("Software Upgrading", ["grad"])
+
+    def test_empty_title(self):
+        """An empty title should not cause an error and should return False."""
+        assert not has_new_grad_signal("", ["new grad"])
+
+    def test_none_title(self):
+        assert not has_new_grad_signal(" ", ["New Grad"])
```
[the remark this answers is not in the document]
Comment (half-measures) [reply tests/test_filter.py]: Ah good catch, fixed.

File: scripts/update_jobs.py [hunk 2]
```diff
@@ -1376,8 +1376,23 @@ def deduplicate_jobs(jobs: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
...
+    #3. Handle edge cases where string itself might be 'nan'
+    elif title.strip().lower() in {'nan', 'none'}:
+        return False
+
+    #Let regex do more lifting, below we check signals in a single pass over job titles
+    combined_signals = "|".join(re.escape(s.lower()) for s in signals)
+    pattern = rf"\b({combined_signals})\b" #Regex pattern to match any signal as a whole word, case-insensitive
+
+    return bool(re.search(pattern, title.lower()))
+
```
[the remark this answers is not in the document]
Comment (half-measures) [reply scripts/update_jobs.py]: I love the arg return thing, added that and cleaned up some of the comments.
